### PR TITLE
Update visual prompting task (finetuning & zero-shot learning)

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] Update data & model pipeline to support multiple prompt types
- [x] Internalize preprocessing pipeline into `learn` and `infer` for standalone module usecase
- [x] Update how to get reference features (saved in the model -> get as input)
    - [x] Save reference info in the specific location to load it for `infer`
- [x] Bug fixes

### How to test

```python
$ tox -vv -e unittest-all-py310-pt2 -- tests/unit/algorithms/visual_prompting
======================================================= 257 passed, 128 warnings in 60.99s (0:01:00) =======================================================
Name                                                                                                                   Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------------------------------------------------
src/otx/algorithms/visual_prompting/adapters/openvino/model_wrappers/openvino_models.py                                   94      3    97%   53, 88, 160    
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/callbacks/inference.py                                     54      0   100%                  
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/config/visual_prompting_config.py                          43      2    95%   108-109        
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/datasets/dataset.py                                       147     18    88%   106, 168, 171, 212, 216, 228-236, 262, 325-331, 401-404
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/datasets/pipelines/sam_transforms.py                       42      0   100%
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/datasets/pipelines/transforms.py                           38      1    97%   59
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/backbones/tiny_vit.py                              333      4    99%   223, 395, 509, 685
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/backbones/vit.py                                   122      0   100%
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/decoders/sam_mask_decoder.py                       153      1    99%   194
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/encoders/sam_image_encoder.py                       10      1    90%   30
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/encoders/sam_prompt_encoder.py                      94      0   100%
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/utils/layer_norm.py                                 14      0   100%
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/utils/mlp_block.py                                  10      0   100%
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py               228      7    97%   219, 229-233, $46-548
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/zero_shot_segment_anything.py     344     65    81%   59-75, 185-186$ 209-213, 322, 324-328, 330-331, 333-334, 341-342, 401, 403, 432-434, 436, 471, 519, 527, 556, 560-561, 587, 589-595, 668, 724, 728-731, 735, 739-743, 747-$62
src/otx/algorithms/visual_prompting/configs/base/configuration.py                                                         40      0   100%
src/otx/algorithms/visual_prompting/configs/base/configuration_enums.py                                                    4      0   100%
src/otx/algorithms/visual_prompting/configs/sam_tiny_vit/configuration.py                                                  4      4     0%   7-13
src/otx/algorithms/visual_prompting/configs/sam_tiny_vit/ptq_optimization_config.py                                        5      5     0%   2-22
src/otx/algorithms/visual_prompting/configs/sam_vit_b/configuration.py                                                     4      4     0%   7-13
src/otx/algorithms/visual_prompting/configs/sam_vit_b/ptq_optimization_config.py                                           5      0   100%
src/otx/algorithms/visual_prompting/configs/zero_shot_sam_tiny_vit/configuration.py                                        4      4     0%   7-13
src/otx/algorithms/visual_prompting/configs/zero_shot_sam_tiny_vit/ptq_optimization_config.py                              5      5     0%   2-22           
src/otx/algorithms/visual_prompting/tasks/inference.py                                                                   326     28    91%   103, 188, 231-232, 237, 353, 363, 368, 395, 417-418, 447, 451-460, 467-468, 472-473, 569, 579, 584, 612, 679                                                               
src/otx/algorithms/visual_prompting/tasks/openvino.py                                                                    550     65    88%   163, 312, 347-348, 356, 404, 410-412, 414, 436, 506, 537, 568, 601, 655, 848, 852-854, 881-882, 886-887, 927, 993, 996, 1016, 1030, 1056, 1065-1067, 1093-1145
src/otx/algorithms/visual_prompting/tasks/train.py                                                                        41      4    90%   90-91, 98-99
src/otx/algorithms/visual_prompting/utils/visual_prompting_utils.py                                                        3      3     0%   17-22

$ tox -vv -e tests-all-py310-pt2 -- tests/integration/cli/visual_prompting
================================================== 23 passed, 4 skipped, 6 warnings in 253.63s (0:04:13) ===================================================

$ tox -vv -e tests-all-py310-pt2 -- tests/e2e/cli/visual_prompting
======================================================== 29 passed, 6 warnings in 470.07s (0:07:50) ========================================================
```

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/visual_prompting/adapters/openvino/model_wrappers/openvino_models.py [hunk 1]
```diff
@@ -94,27 +113,30 @@ def parameters(cls):  # noqa: D102
...
-                    "has_mask_input": np.zeros((1, 1), dtype=np.float32),
-                    "orig_size": np.array(inputs["original_size"], dtype=np.float32).reshape((-1, 2)),
-                    "label": label,
-                }
-            )
+        for prompt_name in ["bboxes", "points"]:
+            for prompt, label in zip(inputs.get(prompt_name), inputs["labels"].get(prompt_name, [])):
+                if prompt_name == "bboxes":
+                    point_coords = self._apply_coords(prompt.reshape(-1, 2, 2), inputs["original_size"])
+                    point_labels = np.array([2, 3], dtype=np.float32).reshape(-1, 2)
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/adapters/openvino/model_wrappers/openvino_models.py:125]: I have two questions.
Why 2x3 for point labels?
Should dtype be np.float32 for labels?

Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/openvino/model_wrappers/openvino_models.py:125]: 1. `point_labels` for bboxes should have both components, 2 is top-left and 3 is bottom-right.
2. Referred to https://github.com/facebookresearch/segment-anything/blob/main/notebooks/onnx_model_example.ipynb. And I tried to use float32 for variables and int64 for size and index due to stability.

File: src/otx/algorithms/visual_prompting/tasks/inference.py [hunk 2]
```diff
@@ -179,7 +179,7 @@ def get_model(config: DictConfig, train_type: TrainType, state_dict: Optional[Or
                         SegmentAnything as VisualPrompter,
                     )
                 elif train_type == TrainType.Zeroshot:
-                    from otx.algorithms.visual_prompting.adapters.pytorch_lightning.models import (
+                    from otx.algorithms.visual_prompting.adapters.pytorch_lightning.models import (  # type: ignore[assignment] # noqa: E501
```
Comment (jaegukhyun) [inline src/otx/algorithms/visual_prompting/tasks/inference.py:182]: Could you explain why we need this noqa?

**Review by jaegukhyun — DISMISSED** [review]
Generally looks fine to me. Do you have plan to implement same logic to v2 branch?

**Comment by sungchul1** [comment]
> Generally looks fine to me. Do you have plan to implement same logic to v2 branch?

Yes, currently working in progress.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:182]: That line becomes longer than 120 due to `# type: ignore[assignment]`.

Comment (jaegukhyun) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:182]: Ah, I want to know why we ignore mypy assignment issue

Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:182]: I think mypy was upset because assigning `VisualPrompter` was written twice, even though if-statement was used.

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py [hunk 3]
```diff
@@ -133,32 +132,27 @@ def set_metrics(self) -> None:
...
+                    "image_encoder.head.bias",
+                ]:
+                    if key in state_dict:
+                        state_dict.pop(key)
 
         if state_dict:
             # state_dict from args.load_from
-            state_dict = replace_state_dict_keys(state_dict, revise_keys)
+            skip_unused_parameters(state_dict)
             self.load_state_dict(state_dict)
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:143]: is there any option to skip unused parameters in `load_state_dict`? maybe `strict=False`?
https://pytorch.org/tutorials/beginner/saving_loading_models.html#id4

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py [hunk 4]
```diff
@@ -172,7 +166,8 @@ def replace_state_dict_keys(state_dict, revise_keys):
                     # load checkpoint from local
                     with open(self.config.model.checkpoint, "rb") as f:
                         state_dict = torch.load(f)
-                state_dict = replace_state_dict_keys(state_dict, revise_keys)
+
+                skip_unused_parameters(state_dict)
                 self.load_state_dict(state_dict, strict=False)
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:157]: Should we have skip_unused_parameters at here?

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 5]
```diff
@@ -149,11 +152,19 @@ def __init__(
...
+        use_point: bool = False,
     ) -> Tuple[Dict[str, Any], Dict[str, Any], List[Dict[str, Any]]]:
         """Pre-process function of OpenVINO Visual Prompting Inferencer for image encoder."""
+        if use_bbox and use_point:
+            logger.warning("If both use_bbox and use_point are set, bboxes and points will be generated randomly.")
+
+        prob = 1.0 if not use_point else 0.0 if not use_bbox and use_point else 0.5
         images, meta = self.model["image_encoder"].preprocess(dataset_item.numpy, extra_processing)
-        prompts = OTXVisualPromptingDataset.get_prompts(dataset_item, self.labels)  # to be replaced
+        prompts = OTXVisualPromptingDataset.get_prompts(dataset_item, self.labels, prob=prob)
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py:167]: I suggest to use other flags than prob in the future.
With a given flag, it looks better to choose the mode by random number generation in get_prompts inside.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 6]
```diff
@@ -338,84 +414,125 @@ def predict(self, dataset_item: DatasetItemEntity) -> List[Annotation]:  # type:
...
+        for i in range(num_iter):
             if i == 0:
                 # First-step prediction
                 mask_input = np.zeros(
                     (1, 1, *map(lambda x: x * 4, inputs["image_embeddings"].shape[2:])), dtype=np.float32
                 )
                 has_mask_input = self.has_mask_inputs[0]
 
-            elif i == 1:
+            elif is_cascade and i == 1:
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py]: we may not need `is_cascade` here.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 7]
```diff
@@ -338,84 +414,125 @@ def predict(self, dataset_item: DatasetItemEntity) -> List[Annotation]:  # type:
...
-                )
+                mask_input, masks = self._postprocess_masks(masks, logits, scores, is_single=True)  # noqa: F821
                 if masks.sum() == 0:
-                    return {"iou_predictions": iou_predictions, "low_res_masks": mask_input}
+                    return {"upscaled_masks": masks}
 
                 has_mask_input = self.has_mask_inputs[1]
 
-            elif i == 2:
+            elif is_cascade and i == 2:
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py]: here too.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 8]
```diff
@@ -338,84 +414,125 @@ def predict(self, dataset_item: DatasetItemEntity) -> List[Annotation]:  # type:
...
+        is_cascade: bool = True,
     ) -> Dict[str, np.ndarray]:
         """Forward function of OpenVINO Visual Prompting Inferencer."""
+        masks: np.ndarray
         logits: np.ndarray
         scores: np.ndarray
-        mask_slice = slice(0, 1)
-        for i in range(3):
+        num_iter = 3 if is_cascade else 1
+        for i in range(num_iter):
```
Comment (wonjuleee) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py:486]: how about dropping for loop here?
do first-step prediction
if is_cascade:
    perform 2 times of post-refinement

**Review by wonjuleee — DISMISSED** [review]
In general, it looks good to me, but I left few minor comments.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py:167]: I'll consider that.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 8 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py:486]: Then, we must repeatedly implement this part.
https://github.com/openvinotoolkit/training_extensions/blob/b3448f67471fcff6e21b52de6d5b22756eef89f7/src/otx/algorithms/visual_prompting/tasks/openvino.py#L520-L527

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:182]:
```python
src/otx/algorithms/visual_prompting/tasks/inference.py:182: error: Incompatible import of "VisualPrompter" (imported name has type "Type[ZeroShotSegmentAnything]", local name has type "Type[SegmentAnything]")  [assignment]
```

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2892/commits/539a37cbd3ec058d70e86704ba159d7aa9f53d26.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 6 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2892/commits/539a37cbd3ec058d70e86704ba159d7aa9f53d26.

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:143]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2892/commits/5c6ffd27cc5088a84a660c52268fcf7471627aa7.

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:157]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2892/commits/5c6ffd27cc5088a84a660c52268fcf7471627aa7.

**Review by wonjuleee — DISMISSED** [review]
LGTM

**Review by jaegukhyun — DISMISSED** [review]

**Review by MarkByun — DISMISSED** [review]
LGTM

**Review by jaegukhyun — APPROVED** [review]

**Review by MarkByun — APPROVED** [review]
